### PR TITLE
shell: Make shell UART backend initialization priority configurable

### LINKED ISSUE #28331: Shell on CDC ACM UART stopped working after PR #24873
I have a custom board that is based on `nucleo_l432kc` with a USB interface. A software that is running a shell
on CDC ACM interface used to work before #24873 got merged.
The last working revision was d92b199173d7cce41f4837cffa93637fe50dfcf2.

**To Reproduce**
Reproducing requires a custom board, however it should be compatible with `nucleo_l432kc` if one manages to connect
a USB cable to it.

1. git clone https://github.com/eugmes/ais-recv-fw.git
2. cd ais-recv-fw.git
3. Change revision in `application/west.yml`
3. west init application
4. west build
5. west flash
6. Connect to the board via CDC ACM serial port using `minicom`, try to press any buttons, observer that there is no visible activity. Note that the USB device is correctly discovered by the OS, and `minicom` does not report any errors. `lsusb -v` output looks fine and no different than the one from a working version.

**Expected behavior**
After connecting to the board via CDC ACM serial port one is able to use shell.

**Impact**
It is not possible to use current versions of Zephyr with the board. Old ones work.

**Environment**
 - OS: MacOS
 - Toolchain: arm-none-eabi-gcc (GNU Arm Embedded Toolchain 9-2020-q2-update) 9.3.1 20200408 (release)
 - Commit SHA: d431b56587a9afbee091180d3511257a148a359f

**Additional context**
I understand that it is hard to diagnose the problem without access to my custom hardware and there my be a problem in my code. I had no success when trying to use USB samples on `stm32l476g_disco` or `stm32f429i_disc1` (before or after the mentioned PR). I may try to use `nucleo_l432kc` if I manage to connect a USB cable to it.

@tbursztyka: maybe you have any ideas what to look at?

### PR DESCRIPTION
**Pull request by eugmes**
Closes: #28331

Configurable initialization priority is useful when using CDC ACM
serial device, for example.

Signed-off-by: Ievgenii Meshcheriakov <ievgenii.meshcheriakov@nordicsemi.no>

### CODE REVIEW

File: subsys/shell/Kconfig.backends [hunk 1]
```diff
@@ -21,6 +21,12 @@ config SHELL_BACKEND_SERIAL
 
 if SHELL_BACKEND_SERIAL
 
+config SHELL_BACKEND_SERIAL_INIT_PRIORITY
```
Comment (jakub-uC) [inline subsys/shell/Kconfig.backends:24]: Shouldn't this parameter be configurable in interrupts priority range?

Comment (eugmes) [reply subsys/shell/Kconfig.backends:24]: I'm not sure what you mean here. This is initialization priority, not interrupt priority. I don't see any bounds applied to other initialization priorities in various Kconfig's.

Comment (nordic-krch) [reply subsys/shell/Kconfig.backends:24]: according to `init.h` it  should be in range 0 to 99.

**Comment by de-nordic** [comment]
Maybe we should create menu where all shell modules containing SYS_INIT would be listed with priority selectable. Not that anyone want to change all of them but as a list that would show that these things start in specified order, or document it somehow.
There seems to be problem with arbitrarily selected priorities: recently some EEPROM devices were initialized before I2C started, work-in-progress feature of FS backed logging has been opening file while being started prior to initialization of file systems, and we all missed it during review.

File: subsys/shell/Kconfig.backends (hunk 1 above)
Comment (eugmes) [reply subsys/shell/Kconfig.backends:24]: I added range clause, but note that it is not used for any other `_INIT_PRIORITY` definition as far as I can see.

Comment (eugmes) [reply subsys/shell/Kconfig.backends:24]: Correction: there is `MODEM_GSM_INIT_PRIORITY` that has the range specified.

**Review by jakub-uC — APPROVED** [review]

**Review by eobalski — CHANGES_REQUESTED** [review]
Blocking for now as this may resolve problem with initialization priority when USB is chosen as shell backend but the priority must be set **manually**.
I'm thinking on a solution to make it **automatic** which will rework shell Kconfig.backend. I believe this would be better way to address the issues with init priorities.

**Comment by jfischer-phytec-iot** [comment]
> Blocking for now as this may resolve problem with initialization priority when USB is chosen as shell backend but the priority must be set **manually**.
> I'm thinking on a solution to make it **automatic** which will rework shell Kconfig.backend. I believe this would be better way to address the issues with init priorities.

I do not think it make sense, I rather would add shell support/configuration to sample/subsys/usb/console like https://github.com/zephyrproject-rtos/zephyr/issues/28331#issuecomment-691278682

**Review by jfischer-no — APPROVED** [review]

**Comment by emob-nordic** [comment]
> I do not think it make sense, I rather would add shell support/configuration to sample/subsys/usb/console like [#28331 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/28331#issuecomment-691278682)

Okay but in that case, the user must be aware of specifying this priority by himself (of course we can add this to example). I'm not quite convinced this is natural way of doing this. I imagined alternate solution like that:

```
choice SHELL_BACKEND
        config SHELL_UART_BACKEND
                
        config SHELL_USB_BACKEND

cofing SHELL_BACKEND_SERIAL_INIT_PRIORITY
        default 70 if SHELL_USB_BACKEND

config DEV_NAME
        default "CDC_ACM_0" if SHELL_USB_BACKEND
        default "UART_0" if SHELL_UART_BACKEND
```

**Comment by emob-nordic** [comment]
I do not insist that hard but I don't remember if there is any other subsystem that requires manually setting init priority for something that the OS offer.

**Comment by emob-nordic** [comment]
> Maybe we should create menu where all shell modules containing SYS_INIT would be listed with priority selectable. Not that anyone want to change all of them but as a list that would show that these things start in specified order, or document it somehow.
> There seems to be problem with arbitrarily selected priorities: recently some EEPROM devices were initialized before I2C started, work-in-progress feature of FS backed logging has been opening file while being started prior to initialization of file systems, and we all missed it during review.

I think this is not strictly related to this particular issue but the Zephyr Kconfig as a whole. We do not have a common mechanism for requesting specific priority for option `A` by enabling option `B`. I can rise an issue for the discussion so we can avoid such problems in the future.

**Comment by jfischer-phytec-iot** [comment]
> config DEV_NAME
>         default "CDC_ACM_0" if SHELL_USB_BACKEND

but it could be CDC_ACM_1 or CDC_ACM_2

**Comment by emob-nordic** [comment]
> > config DEV_NAME
> > default "CDC_ACM_0" if SHELL_USB_BACKEND
> 
> but it could be CDC_ACM_1 or CDC_ACM_2

Yes I know, this is just a draft. I'm thinking if this would be achievable to by more generic options. For now I let others speak as well.

**Comment by eugmes** [comment]
> > Maybe we should create menu where all shell modules containing SYS_INIT would be listed with priority selectable. Not that anyone want to change all of them but as a list that would show that these things start in specified order, or document it somehow.
> > There seems to be problem with arbitrarily selected priorities: recently some EEPROM devices were initialized before I2C started, work-in-progress feature of FS backed logging has been opening file while being started prior to initialization of file systems, and we all missed it during review.
> 
> I think this is not strictly related to this particular issue but the Zephyr Kconfig as a whole. We do not have a common mechanism for requesting specific priority for option `A` by enabling option `B`. You can rise an issue for the discussion so we can avoid such problems in the future.

You may also look at #22545 and #24416.

**Comment by eugmes** [comment]
> I do not insist that hard but I don't remember if there is any other subsystem that requires manually setting init priority for something that the OS offer.

The default initialization priority could just be set to 99 and this will cover all the cases.

**Comment by carlescufi** [comment]
@pabigot could you please review this?

**Comment by de-nordic** [comment]
> 
> The default initialization priority could just be set to 99 and this will cover all the cases.

Maybe not. If you define something that should start after it then you have no wiggle point and would have to modify the priority of this first.

**Comment by emob-nordic** [comment]
> > I do not insist that hard but I don't remember if there is any other subsystem that requires manually setting init priority for something that the OS offer.
> 
> The default initialization priority could just be set to 99 and this will cover all the cases.

If it works with eg. 70 why don't we follow this solution instead of making it configurable? (For USB it works because I've checked that some time ago).

**Review by pabigot — APPROVED** [review]
I agree with @de-nordic: 99 is not a panacea as there's no promise of what will happen with multiple instances at the same priority.  This whole non-device init priority thing has to get fixed, but for now adding another configurable priority seems acceptable.

**Review by de-nordic — APPROVED** [review]
Fixes the problem.

**Review by nordic-krch — APPROVED** [review]

**Review by eobalski — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**